mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -248,25 +235,55 @@ impl Default for BlockHeaderUtxoLoopExtraArgs {
     }
 }
 
-pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
-    weak: UtxoWeak,
-    constructor: impl Fn(UtxoArc) -> T,
+/// This function executes a loop to fetch, validate and store block headers from the connected electrum servers.
+/// sync_status_loop_handle notifies the coin activation function of errors and if the error is temporary or not.
+/// spv_conf is passed from the coin configuration and it determines how headers are validated and stored.
+pub(crate) async fn block_header_utxo_loop(
+    weak: Weak<ElectrumClientImpl>,
     mut sync_status_loop_handle: UtxoSyncStatusLoopHandle,
-    mut block_count: u64,
     spv_conf: SPVConf,
 ) {
-    let args = BlockHeaderUtxoLoopExtraArgs::default();
-    let mut chunk_size = args.chunk_size;
-    while let Some(arc) = weak.upgrade() {
-        let coin = constructor(arc);
-        let ticker = coin.as_ref().conf.ticker.as_str();
-        let client = match &coin.as_ref().rpc_client {
-            UtxoRpcClientEnum::Native(_) => break,
-            UtxoRpcClientEnum::Electrum(client) => client,
+    macro_rules! remove_server_and_break_if_no_servers_left {
+        ($client:expr, $server_address:expr, $ticker:expr, $sync_status_loop_handle:expr) => {
+            if let Err(e) = $client.remove_server($server_address).await {
+                let msg = format!("Error {} on removing server {}!", e, $server_address);
+                // Todo: Permanent error notification should lead to deactivation of coin after applying some fail-safe measures if there are on-going swaps
+                $sync_status_loop_handle.notify_on_permanent_error(msg);
+                break;
+            }
+
+            if $client.is_connections_pool_empty().await {
+                // Todo: Permanent error notification should lead to deactivation of coin after applying some fail-safe measures if there are on-going swaps
+                let msg = format!("All servers are removed for {}!", $ticker);
+                $sync_status_loop_handle.notify_on_permanent_error(msg);
+                break;
+            }
         };
+    }
```

Comment:
why not a function that returns `Some(())` instead of macro...this will be hard to debug I guess.

additionally , this can be a method of `ElectrumClient`